src/FTThresholdServer.cpp

```diff
@@ -0,0 +1,170 @@
+    messageStream << "Force Threshold exceeded!   Threshold: " << forceThreshold
+                  << "   Force: " << mForce.x() << ", " << mForce.y() << ", "
+                  << mForce.z();
+    message = messageStream.str();
+    ROS_WARN(message.c_str());
+  }
+  if (torqueThresholdExceeded) {
+    std::stringstream messageStream;
+    messageStream << "Torque Threshold exceeded!   Threshold: "
+                  << torqueThreshold << "   Torque: " << mTorque.x() << ", "
+                  << mTorque.y() << ", " << mTorque.z();
+    message = messageStream.str();
+    ROS_WARN(message.c_str());
+  }
+
+  return forceThresholdExceeded || torqueThresholdExceeded;
+}
+
+//=============================================================================
+void FTThresholdServer::forceTorqueDataCallback(
+    const geometry_msgs::WrenchStamped &msg) {
+  std::lock_guard<std::mutex> lock(mForceTorqueDataMutex);
+  mForce.x() = msg.wrench.force.x;
+  mForce.y() = msg.wrench.force.y;
+  mForce.z() = msg.wrench.force.z;
+  mTorque.x() = msg.wrench.torque.x;
+  mTorque.y() = msg.wrench.torque.y;
+  mTorque.z() = msg.wrench.torque.z;
+  mTimeOfLastSensorDataReceived = std::chrono::steady_clock::now();
+}
+
+//=============================================================================
+void FTThresholdServer::setForceTorqueThreshold(
+    FTThresholdGoalHandle gh) {
+  const auto goal = gh.getGoal();
+  ROS_INFO_STREAM("Setting thresholds: force = " << goal->force_threshold
+                                                 << ", torque = "
+                                                 << goal->torque_threshold);
+  FTThresholdResult result;
+  result.success = true;
+
+  // check threshold validity
+  if (goal->force_threshold > mForceLimit) {
+    result.success = false;
+    result.message = "Force threshold exceeds maximum sensor value.";
+  } else if (goal->force_threshold < 0.0) {
+    result.success = false;
+    result.message = "Force threshold must be positive or zero.";
+  } else if (goal->torque_threshold > mTorqueLimit) {
+    result.success = false;
+    result.message = "Torque threshold exceeds maximum sensor value.";
+  } else if (goal->torque_threshold < 0.0) {
+    result.success = false;
+    result.message = "Torque threshold must be positive or zero.";
+  }
+
+  if (!result.success) {
+    ROS_WARN_STREAM("FTThresholdServer: " << result.message);
+    gh.setRejected(result);
+    return;
+  }
+
+  gh.setAccepted();
+
+  // Tare if requested
+  if(goal->retare) {
+    // start asynchronous tare request
+    ROS_INFO("Starting Taring");
+    pr_control_msgs::TriggerGoal goal;
+    mTaringCompleted.store(false);
+    // block until tared
+    mTareActionClient->sendGoalAndWait(goal);
+    ROS_INFO("Taring completed!");
+    mTimeOfLastSensorDataReceived = std::chrono::steady_clock::now();
+    mTaringCompleted.store(true);
+  }
+
+  // check that we are not already taring
+  if (!mTaringCompleted.load()) {
```

Comment:
What is this check supposed to do? Why would `mTaringCompleted` be false here? It feels like the only way for this to be the case is that (1) taring wasn't completed previously and (2) the new request didn't ask to retare.

Should this be moved inside the previous branch to prevent taring when a tare is already in progress?

Reply:
I agree that this should be moved before the previous branch to avoid a change in behavior. Not only do we want to avoid double-taring, but we also want to avoid setting a threshold before taring is completed. Basically, if taring is still happening, don't do anything.